### PR TITLE
Code Changes & Turbo/Slider Experimental Functionality

### PR DESCRIPTION
**Pull request by arntsonl**
Lots of code changes, modular inputs and proper add-on naming.

GPAddon will be our base class for now, this will allow other developers to create or modify add-ons.

Turbo pin, LS/RS sliders, turbo shot count default, and turbo LED are only configurable via a uf2 recompile for now, but will be part of the config revamp.

Turbo speed can be changed by holding turbo and pressing UP/DOWN (default is 5 shots per second).

### CODE REVIEW

File: configs/DURAL/BoardConfig.h [hunk 1]
```diff
@@ -26,6 +26,9 @@
 #define PIN_BUTTON_R3   0
 #define PIN_BUTTON_A1   3
 #define PIN_BUTTON_A2   2
+#define PIN_BUTTON_TURBO -1
```
Comment (TheTrainGoes) [inline configs/DURAL/BoardConfig.h:29]: Generally I am of the thought that we should remove the additional configurations in .44.  

We cannot guarantee compatibility until we have tested the code on those specific devices.  I would not want to push an update out and have a bunch of people say their stuff is not working.

I will reach out to the various project owners.

File: configs/DebugBoard/BoardConfig.h [hunk 2]
```diff
@@ -41,6 +41,9 @@
 #define PIN_BUTTON_R3  8
 #define PIN_BUTTON_A1  26
 #define PIN_BUTTON_A2  9 // Not connected
+#define PIN_BUTTON_TURBO -1
```
Comment (TheTrainGoes) [inline configs/DebugBoard/BoardConfig.h:44]: Generally I am of the thought that we should remove the additional configurations in .44.  

We cannot guarantee compatibility until we have tested the code on those specific devices.  I would not want to push an update out and have a bunch of people say their stuff is not working.

I will reach out to the various project owners.

File: configs/Fightboard/BoardConfig.h [hunk 3]
```diff
@@ -33,6 +33,9 @@
 #define PIN_BUTTON_R3   4
 #define PIN_BUTTON_A1   0
 #define PIN_BUTTON_A2   5
+#define PIN_BUTTON_TURBO -1
```
Comment (TheTrainGoes) [inline configs/Fightboard/BoardConfig.h:36]: Generally I am of the thought that we should remove the additional configurations in .44.  

We cannot guarantee compatibility until we have tested the code on those specific devices.  I would not want to push an update out and have a bunch of people say their stuff is not working.

I will reach out to the various project owners.

File: configs/FlatboxRev4/BoardConfig.h [hunk 4]
```diff
@@ -26,6 +26,9 @@
 #define PIN_BUTTON_R3   4
 #define PIN_BUTTON_A1   2
 #define PIN_BUTTON_A2   5
+#define PIN_BUTTON_TURBO -1
```
Comment (TheTrainGoes) [inline configs/FlatboxRev4/BoardConfig.h:29]: Generally I am of the thought that we should remove the additional configurations in .44.  

We cannot guarantee compatibility until we have tested the code on those specific devices.  I would not want to push an update out and have a bunch of people say their stuff is not working.

I will reach out to the various project owners.

File: configs/GeeekPiStick/BoardConfig.h [hunk 5]
```diff
@@ -33,6 +33,10 @@
 #define PIN_BUTTON_A1   21
 #define PIN_BUTTON_A2   22
 
+#define PIN_BUTTON_TURBO -1
```
Comment (TheTrainGoes) [inline configs/GeeekPiStick/BoardConfig.h:36]: Generally I am of the thought that we should remove the additional configurations in .44.  

We cannot guarantee compatibility until we have tested the code on those specific devices.  I would not want to push an update out and have a bunch of people say their stuff is not working.

I will reach out to the various project owners.

File: include/addons/playerleds.h [hunk 6]
```diff
@@ -28,16 +28,15 @@ class PWMPlayerLEDs : public PlayerLEDs
...
-	virtual bool available();  // GPModule
+	virtual bool available();  // GPAddon
 	virtual void setup();
-	virtual void loop();
-	virtual void process(Gamepad *gamepad);
+	virtual void process();
 	virtual std::string name() { return PLEDName; }
-	PLEDModule() : type(PLED_TYPE) { }
-	PLEDModule(PLEDType type) : type(type) { }
+	PlayerLEDAddon() : type(PLED_TYPE) { }
```
Comment (TheTrainGoes) [inline include/addons/playerleds.h:38]: Will connect with you about PLEDs as a whole

File: include/inputs/turbo.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
+#ifndef TURBO_H_
```
Comment (TheTrainGoes) [inline include/inputs/turbo.h:1]: Can we add a little shout out to PicoAnn in here?

File: lib/PlayerLEDs/include/PlayerLEDs.h [hunk 8]
```diff
@@ -4,6 +4,8 @@
 #include <cstring>
 #include <stdint.h>
 
+#include "pico/time.h"
```
Comment (TheTrainGoes) [inline lib/PlayerLEDs/include/PlayerLEDs.h:7]: I am not familiar with this one, lets connect about it.

File: src/inputs/turbo.cpp [hunk 9]
```diff
@@ -0,0 +1,115 @@
+#include "inputs/turbo.h"
```
Comment (TheTrainGoes) [inline src/inputs/turbo.cpp:1]: Sneaky shoutout to PicoAnn in here as well if we can

File: configs/DebugBoard/BoardConfig.h (hunk 2 above)
Comment (TheTrainGoes) [inline configs/DebugBoard/BoardConfig.h:44]: Can we add the Turbo LED pin in here as well with a default -1 and a note that it should be kept on GP25 for the time being.

File: src/inputs/turbo.cpp [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+    if (getMillis() < nextTimer) {
+        return; // don't flip if we haven't reached the next timer
+    }
+
+    bTurboFlicker ^= true; // Button ON/OFF State Reverse
+    nextTimer = getMillis() + uIntervalMS; // interval to flicker-off button
+
+    // Set TURBO LED if a button is going
+    if ( TURBO_LED_PIN != -1 ) {
```
Comment (TheTrainGoes) [inline src/inputs/turbo.cpp:108]: Can we get this added into the config file as a -1 default and a note that it should be kept as GP25 for the time being?

**Review by TheTrainGoes — COMMENTED** [review]
Fantastic work.  Some small comments mostly around configuration.  

I have tested this and it is working without issue.  

Will do some further testing tonight.

File: src/addons/i2cdisplay.cpp [hunk 11]
```diff
@@ -3,67 +3,69 @@
...
-void I2CDisplayModule::setup() {
-	BoardOptions options = Storage::getInstance().getBoardOptions();
+void I2CDisplayAddon::setup() {
+	BoardOptions boardOptions = Storage::getInstance().getBoardOptions();
 	obdI2CInit(&obd,
-	    options.displaySize,
-		options.displayI2CAddress,
-		options.displayFlip,
-		options.displayInvert,
+	    boardOptions.displaySize,
```
Comment (henrebotha) [inline src/addons/i2cdisplay.cpp:20]: May as well fix the indentation while we're here.
```suggestion
		boardOptions.displaySize,
```

File: src/addons/i2cdisplay.cpp [hunk 12]
```diff
@@ -3,67 +3,69 @@
...
-
-		case BUTTON_LAYOUT_HITBOX:
-			drawHitbox(8, 20, 8, 2, gamepad);
-			break;
-
-		case BUTTON_LAYOUT_WASD:
-			drawWasdBox(8, 28, 7, 3, gamepad);
-			break;
+	bool configMode = Storage::getInstance().GetConfigMode();
+	if (configMode == true ) {
```
Comment (henrebotha) [inline src/addons/i2cdisplay.cpp:39]:
```suggestion
	if (configMode == true) {
```

File: src/configmanager.cpp [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+#include "configs/WebConfig.h"
+#include "addons/neopicoleds.h"
+
+void ConfigManager::setup(ConfigType config) {
+	switch(config) {
+		case CONFIG_TYPE_WEB:
+			setupConfig(new WebConfig());
+			break;
+	}
+    this->cType = config;
```
Comment (henrebotha) [inline src/configmanager.cpp:11]: Lot of mixed indentation in this file, let's clean that up.

File: src/gamepad.cpp [hunk 14]
```diff
@@ -42,7 +42,7 @@ void Gamepad::setup()
 	mapButtonR3  = new GamepadButtonMapping(boardOptions.pinButtonR3,  GAMEPAD_MASK_R3);
 	mapButtonA1  = new GamepadButtonMapping(boardOptions.pinButtonA1,  GAMEPAD_MASK_A1);
 	mapButtonA2  = new GamepadButtonMapping(boardOptions.pinButtonA2,  GAMEPAD_MASK_A2);
-
+
```
Comment (henrebotha) [inline src/gamepad.cpp:45]:
```suggestion

```

**Comment by arntsonl** [comment]
Oh yeah we can pull out that version stuff I left commented. And thank you @TheTrainGoes and @henrebotha for looking at this! Indentation fixes look good, I also want to unify the coding style for brackets and camel case as I'm adding on to a coding style I don't always follow.

**Pull request merged**